### PR TITLE
Refactor WaitGroup implementation in tasks.go

### PR DESCRIPTION
**Pull request by shrutiparabgoogle**
Fixes https://github.com/apigee/registry/issues/239

Create a struct for taskQueues
Get rid of the shared variables and use instance specific attributes

### CODE REVIEW

**Comment by shrutiparabgoogle** [comment]
The build will fail because the taskQueue change is yet to be applied to the rest of the commands

**Review by seaneganx — COMMENTED** [review]
Using a lightweight worker function might do just as well with the same amount of work required by the caller. Something like [this example](https://gobyexample.com/worker-pools) combined with [this example](https://gobyexample.com/waitgroups).

Usage would be similar to our current setup:
```
var wg sync.WaitGroup
tasks := make(chan core.Task, 1024)
for i := 0; i < 16; i++ {
	wg.Add(1)
	go core.Worker(ctx, wg, tasks)
}
defer wg.Wait()

// Generate tasks.
tasks <- &computeLintTask{ ... }

// Let workers finish.
close(tasks)
```

We can also consider using a helper to handle most of the setup work. Something like `func Pool(ctx context.Context, n int) (queue chan<- Task, wait func())` should make everything easy for the caller.

```
queue, wait := Pool(ctx, 16)
defer wait()

// Generate tasks.
queue <- &computeLintTask{ ... }
```

**Comment by shrutiparabgoogle** [comment]
The idea was to abstract out the setup and underlying details for creating taskQueues. Hence, I ended up creating the struct. That said, I liked the idea of the helper func which you proposed `func Pool(ctx context.Context, n int) (queue chan<- Task, wait func())` it also doing the same job without the complexity of a struct. Will make changes accordingly.

**Comment by shrutiparabgoogle** [comment]
Update the PR with following changes:
1. Refactored waitgroup implementation
2. Migrated resolve command to use taskQueues and call the sub-commands programmatically. (These changes are mainly reflected in `local-tasks.go`).
3. While importing all the sub-command packages in `local-tasks.go`, there was a cyclic dependency because `upload` command was depending on controller package for `ReadManifestProto`. As follows:
`registry -> resolve -> controller -> upload (in local-tasks.go) -> controller`
Hence, got rid of `manifest-reader.go` and moved the funcs to respective packages.

File: cmd/registry/controller/local-tasks.go [hunk 1]
```diff
@@ -26,16 +38,57 @@ type ExecCommandTask struct {
...
+	case "label":
+		return label.Command(ctx), nil
+	case "list":
+		return list.Command(ctx), nil
+	case "search":
+		return search.Command(ctx), nil
+	case "upload":
+		return upload.Command(ctx), nil
+	case "vocabulary":
+		return vocabulary.Command(ctx), nil
```
Comment (seaneganx) [inline cmd/registry/controller/local-tasks.go:68]: If we ever change these subcommand names, will there be a test that points out this connection is broken?

**Review by seaneganx — COMMENTED** [review]
Would that circular dependency issue still occur if the manifest reader was in its own package? That might be the better approach instead of copying the code into each package that needs it.

File: cmd/registry/core/tasks.go [hunk 2]
```diff
@@ -26,18 +26,29 @@ type Task interface {
 	String() string
 }
 
-var wg sync.WaitGroup
+func WorkerPool(ctx context.Context, n int) (chan Task, func()) {
```
[the remark this answers is not in the document]
Comment (shrutiparabgoogle) [reply cmd/registry/core/tasks.go]: Done.

File: cmd/registry/core/tasks.go [hunk 3]
```diff
@@ -26,18 +26,29 @@ type Task interface {
...
+	var wg sync.WaitGroup
+	taskQueue := make(chan Task, 1024)
+	for i := 0; i < n; i++ {
+		wg.Add(1)
+		go worker(ctx, &wg, taskQueue)
+	}
+	return taskQueue, wg.Wait
```
[the remark this answers is not in the document]
Comment (shrutiparabgoogle) [reply cmd/registry/core/tasks.go]: Done.

File: cmd/registry/controller/controller_test.go [hunk 4]
```diff
@@ -168,6 +171,26 @@ func createUpdateArtifact(
 	}
 }
 
+func readManifestProto(
+	filename string,
+	t *testing.T) (*rpc.Manifest, error) {
```
[the remark this answers is not in the document]
Comment (shrutiparabgoogle) [reply cmd/registry/controller/controller_test.go]: Not relevant anymore.

File: cmd/registry/controller/local-tasks.go (hunk 1 above)
Comment (shrutiparabgoogle) [reply cmd/registry/controller/local-tasks.go:68]: Added another bug to take care to the tests https://github.com/apigee/registry/issues/248

**Comment by shrutiparabgoogle** [comment]
While evaluating the dependency on `manifest_reader` and your suggestion about creating a package outside of controller, I realized the funcs in manifest_reader are very specific to what `resolve.go` wants to do and not very generic to be used anywhere else. Also, `controller_test.go` was using `ReadManifestProto`, which is not actually needed. I replaced that usage with `rpc.Manifest` structs. We were not testing the `ReadManifestProto()` anywhere, so this new approach makes the test error proof as well. Also, I think it improves the time taken to run controller tests, since it doesn't need to do any IO operations any more.

**Review by seaneganx — APPROVED** [review]

File: cmd/registry/cmd/annotate/annotate.go [hunk 5]
```diff
@@ -45,12 +45,8 @@ func Command(ctx context.Context) *cobra.Command {
...
 			}
 
-			taskQueue := make(chan core.Task, 1024)
-			workerCount := 64
-			for i := 0; i < workerCount; i++ {
-				core.WaitGroup().Add(1)
-				go core.Worker(ctx, taskQueue)
-			}
+			taskQueue, closeAndWait := core.WorkerPool(ctx, 64)
+			defer closeAndWait()
```
[the remark this answers is not in the document]
Comment (shrutiparabgoogle) [reply cmd/registry/cmd/annotate/annotate.go]: Done.

**Pull request merged**